src/translation-api.js

```diff
@@ -0,0 +1,107 @@
+import { and, eq } from 'drizzle-orm'
+import { kCreateWithDocId, kSelect } from './datatype/index.js'
+import { hashObject } from './utils.js'
+
+export const ktranslatedLanguageCodeToSchemaNames = Symbol(
+  'translatedLanguageCodeToSchemaNames'
+)
+export default class TranslationApi {
+  /** @type {Map<
+   * import('@mapeo/schema').TranslationValue['languageCode'],
+   * Set<import('@mapeo/schema/dist/types.js').SchemaName>>} */
+  #translatedLanguageCodeToSchemaNames = new Map()
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').translationTable,
+   *   'translation',
+   *   import('@mapeo/schema').Translation,
+   *   import('@mapeo/schema').TranslationValue
+   * >}  opts.dataType
+   * @param {typeof import('./schema/project.js').translationTable} opts.table
+   */
+  constructor({ dataType, table }) {
+    this.dataType = dataType
+    this.table = table
+  }
+
+  /**
+   * @param {import('@mapeo/schema').TranslationValue} value
+   */
+  async put(value) {
+    /* eslint-disable no-unused-vars */
+    const { message, ...identifiers } = value
+    const docId = hashObject(identifiers)
+    let existing
+    try {
+      existing = await this.dataType.getByDocId(docId)
+      if (existing) {
+        await this.dataType.update(existing.versionId, value)
+      }
+    } catch (e) {
+      existing = await this.dataType[kCreateWithDocId](docId, value)
+    }
+    return existing.docId
+  }
+
+  /**
+   * @param {import('type-fest').SetOptional<
+   * Omit<import('@mapeo/schema').TranslationValue,'schemaName' | 'message'>,
+   * 'fieldRef' | 'regionCode'>} value - We omit the schemaName and message, and we make fieldRef and RegionCode optional
+   */
+  async get(value) {
+    const docTypeIsTranslatedToLanguage =
+      this.#translatedLanguageCodeToSchemaNames
+        .get(value.languageCode)
```

Comment:
I think this will be empty on app startup, which means that we'll never be able to get docs that were added on a different launch. For example:

1. I start the app and add a translation, which gets indexed into the cache.
2. I restart the app. The cache is wiped.
3. I call `TranslationApi.prototype.get`. Because the cache is empty, we won't look up the document.

What if we just removed the cache entirely?

Reply:
Yeah, what you say makes sense. I'm wondering how hard would it be to populate the cache when restarting the app. I _think_ it would be smth like calling `this.#datatype.getMany()` on the `TranslationApi` constructor and using that to populate the cache?  I would need to be able to set up tests differently to allow running them with a persistent db - so that I can test this specific thing - . Doing that may be a little more involved that what I'm thinking. Buut it may be worth it. wdyt?

Reply:
Is the cache primarily for performance? If so: how slow is it without the cache?

I wonder if we can remove it because the "no cache" performance is fine.

Reply:
I did a quick (and pretty non-exhaustive) perf test where we query 150_000 docs with non-existent `languageCodes`. With the cache it runs in 1.2 secs and without it it runs in 15.36 secs. 
On the other side I'm wondering how many combinations of `languageCode` and `schemaName` a sole user would have to justify having a cache.
I wonder what @gmaclennan thinks.

Reply:
I wonder if you could recoup some of that performance with a SQLite index...

Reply:
The idea of the cache is for read performance of records that _do not have_ translations. Currently the only docs that might have translations are presets and fields, and there will never be many of these. So the performance issue we are worried about is looking up and not finding translations for observations really. The index is minimal - I don't think it will ever be large - so maybe we could just store it as json somewhere, maybe in SQLite? It won't change much and will only need read once on app startup.

Reply:
In any reason, we would want to test the cache right? Which means having some sort of persistence on subsequent runs of a test? I think I'm going to need help with that.

Reply:
> If there's valid reasons for keeping the cache, and its going to be pretty minimal, shouldn't rebuilding it from the saved cores when the app starts be enough??

To rebuild it from saved cores, then we need to re-index all of the cores, which is an expensive operation. I think we need to persist it somehow, interested in @EvanHahn 's thoughts though.

Reply:
> In any reason, we would want to test the cache right? Which means having some sort of persistence on subsequent runs of a test? I think I'm going to need help with that.

Yes, and we actually don't have enough e2e tests that include closing the instance and then re-opening it. Maybe @EvanHahn can help with setting up some helpers for that.

Reply:
[this PR](https://github.com/digidem/mapeo-core-next/pull/580/files) adds a helper to have a persistent project between runs, until it is merged I commented the test and the population of the cache when instancing the `TranslationApi`([here](https://github.com/digidem/mapeo-core-next/blob/06da1a0cccab406cc10434c44ffe197b14fae428/src/translation-api.js#L30)). 
Regarding saving the cache on a JSON or creating an index in SQL lite, it may be worth it, but maybe not? (I should test how much time it takes to re-build the cache, but we know that this cache is not going to be pretty big)